File: julia/Tensor4all.jl/src/QuanticsGrids.jl
"""
    QuanticsGrids

Quantics grid types for efficient conversion between quantics indices,
grid indices, and original coordinates.

Provides two grid types:
- `DiscretizedGrid`: continuous domain with floating-point coordinates
- `InherentDiscreteGrid`: integer coordinate domain

# Example

```julia
using Tensor4all.QuanticsGrids

# Create a 1D grid with 8 bits (256 grid points) over [0, 1)
grid = DiscretizedGrid(1, 8, [0.0], [1.0])

# Convert coordinates
quantics = origcoord_to_quantics(grid, [0.5])
coord = quantics_to_origcoord(grid, quantics)
```
"""
module QuanticsGrids

using ..Tensor4all: C_API

export DiscretizedGrid, InherentDiscreteGrid
export origcoord_to_quantics, quantics_to_origcoord
export origcoord_to_grididx, grididx_to_origcoord
export grididx_to_quantics, quantics_to_grididx

# ============================================================================
# Unfolding scheme helper
# ============================================================================

function _unfolding_to_cint(unfolding::Symbol)
    if unfolding == :fused
        return C_API.UNFOLDING_FUSED
    elseif unfolding == :interleaved
        return C_API.UNFOLDING_INTERLEAVED
    else
        error("Unknown unfolding scheme: $unfolding. Use :fused or :interleaved.")
    end
end

# ============================================================================
# DiscretizedGrid
# ============================================================================

"""
    DiscretizedGrid

A discretized grid with continuous domain and floating-point coordinates.

# Constructors

- `DiscretizedGrid(ndims, R, lower, upper; unfolding=:fused)` - uniform R
- `DiscretizedGrid(ndims, rs, lower, upper; unfolding=:fused)` - per-dimension R

# Properties

- `ndims(g)` - Number of dimensions
- `rs(g)` - Resolution (bits) per dimension
- `local_dimensions(g)` - Local dimensions of tensor sites
- `lower_bound(g)` - Lower bounds per dimension
- `upper_bound(g)` - Upper bounds per dimension
- `grid_step(g)` - Grid spacing per dimension
"""
mutable struct DiscretizedGrid
    ptr::Ptr{Cvoid}
    _ndims::Int
    _rs::Vector{Int}

    function DiscretizedGrid(ptr::Ptr{Cvoid}, ndims::Int, rs::Vector{Int})
        if ptr == C_NULL
            error("Failed to create DiscretizedGrid (null pointer from C API)")
        end
        grid = new(ptr, ndims, rs)
        finalizer(grid) do x
            C_API.t4a_qgrid_disc_release(x.ptr)
        end
        return grid
    end
end

"""
    DiscretizedGrid(ndims::Int, R::Int, lower, upper; unfolding=:fused)

Create a DiscretizedGrid with uniform resolution R for all dimensions.
"""
function DiscretizedGrid(ndims::Int, R::Int, lower::AbstractVector{<:Real},
                         upper::AbstractVector{<:Real}; unfolding::Symbol=:fused)
    rs_vec = fill(R, ndims)
    return DiscretizedGrid(ndims, rs_vec, lower, upper; unfolding=unfolding)
end

"""
    DiscretizedGrid(ndims::Int, rs::Vector{Int}, lower, upper; unfolding=:fused)

Create a DiscretizedGrid with per-dimension resolutions.
"""
function DiscretizedGrid(ndims::Int, rs::AbstractVector{<:Integer},
                         lower::AbstractVector{<:Real}, upper::AbstractVector{<:Real};
                         unfolding::Symbol=:fused)
    length(rs) == ndims || error("rs must have length $ndims, got $(length(rs))")
    length(lower) == ndims || error("lower must have length $ndims, got $(length(lower))")
    length(upper) == ndims || error("upper must have length $ndims, got $(length(upper))")

    rs_c = Csize_t.(rs)
    lower_c = Cdouble.(lower)
    upper_c = Cdouble.(upper)
    unfolding_c = _unfolding_to_cint(unfolding)

    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_qgrid_disc_new(ndims, rs_c, lower_c, upper_c, unfolding_c, out)
    C_API.check_status(status)

    return DiscretizedGrid(out[], ndims, Int.(rs))
end

# Properties

"""
    ndims(g::DiscretizedGrid) -> Int

Number of dimensions.
"""
Base.ndims(g::DiscretizedGrid) = g._ndims

"""
    rs(g::DiscretizedGrid) -> Vector{Int}

Resolution (bits) per dimension.
"""
rs(g::DiscretizedGrid) = g._rs

"""
    local_dimensions(g::DiscretizedGrid) -> Vector{Int}

Local dimensions of all tensor sites.
"""
function local_dimensions(g::DiscretizedGrid)
    max_sites = sum(g._rs) + g._ndims  # upper bound
    out = Vector{Csize_t}(undef, max_sites)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_disc_local_dims(g.ptr, out, max_sites, n_out)
    C_API.check_status(status)
    return Int.(out[1:n_out[]])
end

"""
    lower_bound(g::DiscretizedGrid) -> Vector{Float64}

Lower bounds per dimension.
"""
function lower_bound(g::DiscretizedGrid)
    out = Vector{Cdouble}(undef, g._ndims)
    status = C_API.t4a_qgrid_disc_lower_bound(g.ptr, out, g._ndims)
    C_API.check_status(status)
    return out
end

"""
    upper_bound(g::DiscretizedGrid) -> Vector{Float64}

Upper bounds per dimension.
"""
function upper_bound(g::DiscretizedGrid)
    out = Vector{Cdouble}(undef, g._ndims)
    status = C_API.t4a_qgrid_disc_upper_bound(g.ptr, out, g._ndims)
    C_API.check_status(status)
    return out
end

"""
    grid_step(g::DiscretizedGrid) -> Vector{Float64}

Grid spacing per dimension.
"""
function grid_step(g::DiscretizedGrid)
    out = Vector{Cdouble}(undef, g._ndims)
    status = C_API.t4a_qgrid_disc_grid_step(g.ptr, out, g._ndims)
    C_API.check_status(status)
    return out
end

# Coordinate conversions

"""
    origcoord_to_quantics(g::DiscretizedGrid, coord) -> Vector{Int64}

Convert original (continuous) coordinates to quantics indices.
"""
function origcoord_to_quantics(g::DiscretizedGrid, coord::AbstractVector{<:Real})
    coord_c = Cdouble.(coord)
    max_sites = sum(g._rs) + g._ndims
    out = Vector{Int64}(undef, max_sites)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_disc_origcoord_to_quantics(
        g.ptr, coord_c, length(coord), out, max_sites, n_out)
    C_API.check_status(status)
    return out[1:n_out[]]
end

"""
    quantics_to_origcoord(g::DiscretizedGrid, quantics) -> Vector{Float64}

Convert quantics indices to original (continuous) coordinates.
"""
function quantics_to_origcoord(g::DiscretizedGrid, quantics::AbstractVector{<:Integer})
    quantics_c = Int64.(quantics)
    out = Vector{Cdouble}(undef, g._ndims)
    status = C_API.t4a_qgrid_disc_quantics_to_origcoord(
        g.ptr, quantics_c, length(quantics), out, g._ndims)
    C_API.check_status(status)
    return out
end

"""
    origcoord_to_grididx(g::DiscretizedGrid, coord) -> Vector{Int64}

Convert original coordinates to grid indices (1-indexed).
"""
function origcoord_to_grididx(g::DiscretizedGrid, coord::AbstractVector{<:Real})
    coord_c = Cdouble.(coord)
    out = Vector{Int64}(undef, g._ndims)
    status = C_API.t4a_qgrid_disc_origcoord_to_grididx(
        g.ptr, coord_c, length(coord), out, g._ndims)
    C_API.check_status(status)
    return out
end

"""
    grididx_to_origcoord(g::DiscretizedGrid, grididx) -> Vector{Float64}

Convert grid indices (1-indexed) to original coordinates.
"""
function grididx_to_origcoord(g::DiscretizedGrid, grididx::AbstractVector{<:Integer})
    grididx_c = Int64.(grididx)
    out = Vector{Cdouble}(undef, g._ndims)
    status = C_API.t4a_qgrid_disc_grididx_to_origcoord(
        g.ptr, grididx_c, length(grididx), out, g._ndims)
    C_API.check_status(status)
    return out
end

"""
    grididx_to_quantics(g::DiscretizedGrid, grididx) -> Vector{Int64}

Convert grid indices to quantics indices.
"""
function grididx_to_quantics(g::DiscretizedGrid, grididx::AbstractVector{<:Integer})
    grididx_c = Int64.(grididx)
    max_sites = sum(g._rs) + g._ndims
    out = Vector{Int64}(undef, max_sites)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_disc_grididx_to_quantics(
        g.ptr, grididx_c, length(grididx), out, max_sites, n_out)
    C_API.check_status(status)
    return out[1:n_out[]]
end

"""
    quantics_to_grididx(g::DiscretizedGrid, quantics) -> Vector{Int64}

Convert quantics indices to grid indices.
"""
function quantics_to_grididx(g::DiscretizedGrid, quantics::AbstractVector{<:Integer})
    quantics_c = Int64.(quantics)
    out = Vector{Int64}(undef, g._ndims)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_disc_quantics_to_grididx(
        g.ptr, quantics_c, length(quantics), out, g._ndims, n_out)
    C_API.check_status(status)
    return out[1:n_out[]]
end

# Display
function Base.show(io::IO, g::DiscretizedGrid)
    print(io, "DiscretizedGrid(ndims=$(g._ndims), rs=$(g._rs))")
end

# ============================================================================
# InherentDiscreteGrid
# ============================================================================

"""
    InherentDiscreteGrid

A discrete grid for quantics tensor train representations with integer coordinates.

# Constructors

- `InherentDiscreteGrid(ndims, R; origin=nothing, unfolding=:fused)` - uniform R
- `InherentDiscreteGrid(ndims, rs; origin=nothing, unfolding=:fused)` - per-dimension R
"""
mutable struct InherentDiscreteGrid
    ptr::Ptr{Cvoid}
    _ndims::Int
    _rs::Vector{Int}

    function InherentDiscreteGrid(ptr::Ptr{Cvoid}, ndims::Int, rs::Vector{Int})
        if ptr == C_NULL
            error("Failed to create InherentDiscreteGrid (null pointer from C API)")
        end
        grid = new(ptr, ndims, rs)
        finalizer(grid) do x
            C_API.t4a_qgrid_int_release(x.ptr)
        end
        return grid
    end
end

"""
    InherentDiscreteGrid(ndims::Int, R::Int; origin=nothing, unfolding=:fused)

Create an InherentDiscreteGrid with uniform resolution R.
"""
function InherentDiscreteGrid(ndims::Int, R::Int;
                               origin::Union{Nothing,AbstractVector{<:Integer}}=nothing,
                               unfolding::Symbol=:fused)
    rs_vec = fill(R, ndims)
    return InherentDiscreteGrid(ndims, rs_vec; origin=origin, unfolding=unfolding)
end

"""
    InherentDiscreteGrid(ndims::Int, rs::Vector{Int}; origin=nothing, unfolding=:fused)

Create an InherentDiscreteGrid with per-dimension resolutions.
"""
function InherentDiscreteGrid(ndims::Int, rs::AbstractVector{<:Integer};
                               origin::Union{Nothing,AbstractVector{<:Integer}}=nothing,
                               unfolding::Symbol=:fused)
    length(rs) == ndims || error("rs must have length $ndims, got $(length(rs))")

    rs_c = Csize_t.(rs)
    unfolding_c = _unfolding_to_cint(unfolding)

    origin_ptr = if origin !== nothing
        length(origin) == ndims || error("origin must have length $ndims, got $(length(origin))")
        Int64.(origin)
    else
        C_NULL
    end

    out = Ref{Ptr{Cvoid}}(C_NULL)
    status = C_API.t4a_qgrid_int_new(ndims, rs_c, origin_ptr, unfolding_c, out)
    C_API.check_status(status)

    return InherentDiscreteGrid(out[], ndims, Int.(rs))
end

# Properties

Base.ndims(g::InherentDiscreteGrid) = g._ndims
rs(g::InherentDiscreteGrid) = g._rs

function local_dimensions(g::InherentDiscreteGrid)
    max_sites = sum(g._rs) + g._ndims
    out = Vector{Csize_t}(undef, max_sites)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_int_local_dims(g.ptr, out, max_sites, n_out)
    C_API.check_status(status)
    return Int.(out[1:n_out[]])
end

"""
    origin(g::InherentDiscreteGrid) -> Vector{Int64}

Origin per dimension.
"""
function origin(g::InherentDiscreteGrid)
    out = Vector{Int64}(undef, g._ndims)
    status = C_API.t4a_qgrid_int_origin(g.ptr, out, g._ndims)
    C_API.check_status(status)
    return out
end

# Coordinate conversions

function origcoord_to_quantics(g::InherentDiscreteGrid, coord::AbstractVector{<:Integer})
    coord_c = Int64.(coord)
    max_sites = sum(g._rs) + g._ndims
    out = Vector{Int64}(undef, max_sites)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_int_origcoord_to_quantics(
        g.ptr, coord_c, length(coord), out, max_sites, n_out)
    C_API.check_status(status)
    return out[1:n_out[]]
end

function quantics_to_origcoord(g::InherentDiscreteGrid, quantics::AbstractVector{<:Integer})
    quantics_c = Int64.(quantics)
    out = Vector{Int64}(undef, g._ndims)
    status = C_API.t4a_qgrid_int_quantics_to_origcoord(
        g.ptr, quantics_c, length(quantics), out, g._ndims)
    C_API.check_status(status)
    return out
end

function origcoord_to_grididx(g::InherentDiscreteGrid, coord::AbstractVector{<:Integer})
    coord_c = Int64.(coord)
    out = Vector{Int64}(undef, g._ndims)
    status = C_API.t4a_qgrid_int_origcoord_to_grididx(
        g.ptr, coord_c, length(coord), out, g._ndims)
    C_API.check_status(status)
    return out
end

function grididx_to_origcoord(g::InherentDiscreteGrid, grididx::AbstractVector{<:Integer})
    grididx_c = Int64.(grididx)
    out = Vector{Int64}(undef, g._ndims)
    status = C_API.t4a_qgrid_int_grididx_to_origcoord(
        g.ptr, grididx_c, length(grididx), out, g._ndims)
    C_API.check_status(status)
    return out
end

function grididx_to_quantics(g::InherentDiscreteGrid, grididx::AbstractVector{<:Integer})
    grididx_c = Int64.(grididx)
    max_sites = sum(g._rs) + g._ndims
    out = Vector{Int64}(undef, max_sites)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_int_grididx_to_quantics(
        g.ptr, grididx_c, length(grididx), out, max_sites, n_out)
    C_API.check_status(status)
    return out[1:n_out[]]
end

function quantics_to_grididx(g::InherentDiscreteGrid, quantics::AbstractVector{<:Integer})
    quantics_c = Int64.(quantics)
    out = Vector{Int64}(undef, g._ndims)
    n_out = Ref{Csize_t}(0)
    status = C_API.t4a_qgrid_int_quantics_to_grididx(
        g.ptr, quantics_c, length(quantics), out, g._ndims, n_out)
    C_API.check_status(status)
    return out[1:n_out[]]
end

# Display
function Base.show(io::IO, g::InherentDiscreteGrid)
    print(io, "InherentDiscreteGrid(ndims=$(g._ndims), rs=$(g._rs))")
end

end # module QuanticsGrids

File: python/tensor4all/src/tensor4all/quanticsgrids.py
"""QuanticsGrids - Quantics grid types for coordinate conversions.

This module provides two grid types for quantics tensor train representations:
- ``DiscretizedGrid``: continuous domain with floating-point coordinates
- ``InherentDiscreteGrid``: integer coordinate domain

Both support coordinate conversions between original coordinates, grid indices,
and quantics indices.

Examples
--------
>>> from tensor4all.quanticsgrids import DiscretizedGrid
>>>
>>> # Create a 1D grid with 8 bits (256 grid points) over [0, 1)
>>> grid = DiscretizedGrid(1, 8, [0.0], [1.0])
>>>
>>> # Convert coordinates
>>> quantics = grid.origcoord_to_quantics([0.5])
>>> coord = grid.quantics_to_origcoord(quantics)
"""

from __future__ import annotations

from typing import Sequence, Optional, List

from ._ffi import ffi
from ._capi import get_lib, check_status


def _unfolding_to_int(unfolding: str) -> int:
    """Convert unfolding scheme string to C enum value."""
    if unfolding == "fused":
        return 0
    elif unfolding == "interleaved":
        return 1
    else:
        raise ValueError(f"Unknown unfolding scheme: {unfolding!r}. Use 'fused' or 'interleaved'.")


class DiscretizedGrid:
    """A discretized grid with continuous domain and floating-point coordinates.

    Parameters
    ----------
    ndims : int
        Number of dimensions.
    R : int or Sequence[int]
        Resolution (bits) per dimension. If int, same for all dimensions.
    lower : Sequence[float]
        Lower bounds per dimension.
    upper : Sequence[float]
        Upper bounds per dimension.
    unfolding : str, optional
        Unfolding scheme: "fused" (default) or "interleaved".

    Examples
    --------
    >>> grid = DiscretizedGrid(1, 3, [0.0], [1.0])
    >>> grid.ndims
    1
    >>> grid.rs
    [3]
    >>> grid.grid_step
    [0.125]
    """

    def __init__(
        self,
        ndims: int,
        R,
        lower: Sequence[float],
        upper: Sequence[float],
        unfolding: str = "fused",
    ):
        lib = get_lib()

        if isinstance(R, int):
            rs_list = [R] * ndims
        else:
            rs_list = list(R)
            if len(rs_list) != ndims:
                raise ValueError(f"rs must have length {ndims}, got {len(rs_list)}")

        if len(lower) != ndims:
            raise ValueError(f"lower must have length {ndims}, got {len(lower)}")
        if len(upper) != ndims:
            raise ValueError(f"upper must have length {ndims}, got {len(upper)}")

        rs_c = ffi.new("size_t[]", rs_list)
        lower_c = ffi.new("double[]", list(lower))
        upper_c = ffi.new("double[]", list(upper))
        out = ffi.new("t4a_qgrid_disc**")

        status = lib.t4a_qgrid_disc_new(
            ndims, rs_c, lower_c, upper_c, _unfolding_to_int(unfolding), out
        )
        check_status(status, "DiscretizedGrid.__init__")

        self._ptr = out[0]
        self._ndims = ndims
        self._rs = rs_list

    def __del__(self):
        if hasattr(self, "_ptr") and self._ptr != ffi.NULL:
            get_lib().t4a_qgrid_disc_release(self._ptr)
            self._ptr = ffi.NULL

    @property
    def ndims(self) -> int:
        """Number of dimensions."""
        return self._ndims

    @property
    def rs(self) -> List[int]:
        """Resolution (bits) per dimension."""
        return list(self._rs)

    @property
    def local_dimensions(self) -> List[int]:
        """Local dimensions of all tensor sites."""
        max_sites = sum(self._rs) + self._ndims
        out = ffi.new("size_t[]", max_sites)
        n_out = ffi.new("size_t*")
        status = get_lib().t4a_qgrid_disc_local_dims(self._ptr, out, max_sites, n_out)
        check_status(status, "local_dimensions")
        return [out[i] for i in range(n_out[0])]

    @property
    def lower_bound(self) -> List[float]:
        """Lower bounds per dimension."""
        out = ffi.new("double[]", self._ndims)
        status = get_lib().t4a_qgrid_disc_lower_bound(self._ptr, out, self._ndims)
        check_status(status, "lower_bound")
        return [out[i] for i in range(self._ndims)]

    @property
    def upper_bound(self) -> List[float]:
        """Upper bounds per dimension."""
        out = ffi.new("double[]", self._ndims)
        status = get_lib().t4a_qgrid_disc_upper_bound(self._ptr, out, self._ndims)
        check_status(status, "upper_bound")
        return [out[i] for i in range(self._ndims)]

    @property
    def grid_step(self) -> List[float]:
        """Grid spacing per dimension."""
        out = ffi.new("double[]", self._ndims)
        status = get_lib().t4a_qgrid_disc_grid_step(self._ptr, out, self._ndims)
        check_status(status, "grid_step")
        return [out[i] for i in range(self._ndims)]

    def origcoord_to_quantics(self, coord: Sequence[float]) -> List[int]:
        """Convert original (continuous) coordinates to quantics indices.

        Parameters
        ----------
        coord : Sequence[float]
            Original coordinates, one per dimension.

        Returns
        -------
        List[int]
            Quantics indices (1-indexed).
        """
        lib = get_lib()
        coord_c = ffi.new("double[]", list(coord))
        max_sites = sum(self._rs) + self._ndims
        out = ffi.new("int64_t[]", max_sites)
        n_out = ffi.new("size_t*")
        status = lib.t4a_qgrid_disc_origcoord_to_quantics(
            self._ptr, coord_c, len(coord), out, max_sites, n_out
        )
        check_status(status, "origcoord_to_quantics")
        return [out[i] for i in range(n_out[0])]

    def quantics_to_origcoord(self, quantics: Sequence[int]) -> List[float]:
        """Convert quantics indices to original (continuous) coordinates.

        Parameters
        ----------
        quantics : Sequence[int]
            Quantics indices (1-indexed).

        Returns
        -------
        List[float]
            Original coordinates.
        """
        lib = get_lib()
        quantics_c = ffi.new("int64_t[]", list(quantics))
        out = ffi.new("double[]", self._ndims)
        status = lib.t4a_qgrid_disc_quantics_to_origcoord(
            self._ptr, quantics_c, len(quantics), out, self._ndims
        )
        check_status(status, "quantics_to_origcoord")
        return [out[i] for i in range(self._ndims)]

    def origcoord_to_grididx(self, coord: Sequence[float]) -> List[int]:
        """Convert original coordinates to grid indices (1-indexed).

        Parameters
        ----------
        coord : Sequence[float]
            Original coordinates.

        Returns
        -------
        List[int]
            Grid indices (1-indexed).
        """
        lib = get_lib()
        coord_c = ffi.new("double[]", list(coord))
        out = ffi.new("int64_t[]", self._ndims)
        status = lib.t4a_qgrid_disc_origcoord_to_grididx(
            self._ptr, coord_c, len(coord), out, self._ndims
        )
        check_status(status, "origcoord_to_grididx")
        return [out[i] for i in range(self._ndims)]

    def grididx_to_origcoord(self, grididx: Sequence[int]) -> List[float]:
        """Convert grid indices (1-indexed) to original coordinates.

        Parameters
        ----------
        grididx : Sequence[int]
            Grid indices (1-indexed).

        Returns
        -------
        List[float]
            Original coordinates.
        """
        lib = get_lib()
        grididx_c = ffi.new("int64_t[]", list(grididx))
        out = ffi.new("double[]", self._ndims)
        status = lib.t4a_qgrid_disc_grididx_to_origcoord(
            self._ptr, grididx_c, len(grididx), out, self._ndims
        )
        check_status(status, "grididx_to_origcoord")
        return [out[i] for i in range(self._ndims)]

    def grididx_to_quantics(self, grididx: Sequence[int]) -> List[int]:
        """Convert grid indices to quantics indices.

        Parameters
        ----------
        grididx : Sequence[int]
            Grid indices (1-indexed).

        Returns
        -------
        List[int]
            Quantics indices (1-indexed).
        """
        lib = get_lib()
        grididx_c = ffi.new("int64_t[]", list(grididx))
        max_sites = sum(self._rs) + self._ndims
        out = ffi.new("int64_t[]", max_sites)
        n_out = ffi.new("size_t*")
        status = lib.t4a_qgrid_disc_grididx_to_quantics(
            self._ptr, grididx_c, len(grididx), out, max_sites, n_out
        )
        check_status(status, "grididx_to_quantics")
        return [out[i] for i in range(n_out[0])]

    def quantics_to_grididx(self, quantics: Sequence[int]) -> List[int]:
        """Convert quantics indices to grid indices.

        Parameters
        ----------
        quantics : Sequence[int]
            Quantics indices (1-indexed).

        Returns
        -------
        List[int]
            Grid indices (1-indexed).
        """
        lib = get_lib()
        quantics_c = ffi.new("int64_t[]", list(quantics))
        out = ffi.new("int64_t[]", self._ndims)
        n_out = ffi.new("size_t*")
        status = lib.t4a_qgrid_disc_quantics_to_grididx(
            self._ptr, quantics_c, len(quantics), out, self._ndims, n_out
        )
        check_status(status, "quantics_to_grididx")
        return [out[i] for i in range(n_out[0])]

    def __repr__(self) -> str:
        return f"DiscretizedGrid(ndims={self._ndims}, rs={self._rs})"


class InherentDiscreteGrid:
    """A discrete grid for quantics tensor train representations with integer coordinates.

    Parameters
    ----------
    ndims : int
        Number of dimensions.
    R : int or Sequence[int]
        Resolution (bits) per dimension. If int, same for all dimensions.
    origin : Sequence[int] or None, optional
        Origin per dimension (default: all 1s).
    unfolding : str, optional
        Unfolding scheme: "fused" (default) or "interleaved".

    Examples
    --------
    >>> grid = InherentDiscreteGrid(1, 3)
    >>> grid.ndims
    1
    >>> grid.rs
    [3]
    """

    def __init__(
        self,
        ndims: int,
        R,
        origin: Optional[Sequence[int]] = None,
        unfolding: str = "fused",
    ):
        lib = get_lib()

        if isinstance(R, int):
            rs_list = [R] * ndims
        else:
            rs_list = list(R)
            if len(rs_list) != ndims:
                raise ValueError(f"rs must have length {ndims}, got {len(rs_list)}")

        rs_c = ffi.new("size_t[]", rs_list)

        if origin is not None:
            if len(origin) != ndims:
                raise ValueError(f"origin must have length {ndims}, got {len(origin)}")
            origin_c = ffi.new("int64_t[]", list(origin))
        else:
            origin_c = ffi.NULL

        out = ffi.new("t4a_qgrid_int**")
        status = lib.t4a_qgrid_int_new(
            ndims, rs_c, origin_c, _unfolding_to_int(unfolding), out
        )
        check_status(status, "InherentDiscreteGrid.__init__")

        self._ptr = out[0]
        self._ndims = ndims
        self._rs = rs_list

    def __del__(self):
        if hasattr(self, "_ptr") and self._ptr != ffi.NULL:
            get_lib().t4a_qgrid_int_release(self._ptr)
            self._ptr = ffi.NULL

    @property
    def ndims(self) -> int:
        """Number of dimensions."""
        return self._ndims

    @property
    def rs(self) -> List[int]:
        """Resolution (bits) per dimension."""
        return list(self._rs)

    @property
    def local_dimensions(self) -> List[int]:
        """Local dimensions of all tensor sites."""
        max_sites = sum(self._rs) + self._ndims
        out = ffi.new("size_t[]", max_sites)
        n_out = ffi.new("size_t*")
        status = get_lib().t4a_qgrid_int_local_dims(self._ptr, out, max_sites, n_out)
        check_status(status, "local_dimensions")
        return [out[i] for i in range(n_out[0])]

    @property
    def origin(self) -> List[int]:
        """Origin per dimension."""
        out = ffi.new("int64_t[]", self._ndims)
        status = get_lib().t4a_qgrid_int_origin(self._ptr, out, self._ndims)
        check_status(status, "origin")
        return [out[i] for i in range(self._ndims)]

    def origcoord_to_quantics(self, coord: Sequence[int]) -> List[int]:
        """Convert original (integer) coordinates to quantics indices."""
        lib = get_lib()
        coord_c = ffi.new("int64_t[]", list(coord))
        max_sites = sum(self._rs) + self._ndims
        out = ffi.new("int64_t[]", max_sites)
        n_out = ffi.new("size_t*")
        status = lib.t4a_qgrid_int_origcoord_to_quantics(
            self._ptr, coord_c, len(coord), out, max_sites, n_out
        )
        check_status(status, "origcoord_to_quantics")
        return [out[i] for i in range(n_out[0])]

    def quantics_to_origcoord(self, quantics: Sequence[int]) -> List[int]:
        """Convert quantics indices to original (integer) coordinates."""
        lib = get_lib()
        quantics_c = ffi.new("int64_t[]", list(quantics))
        out = ffi.new("int64_t[]", self._ndims)
        status = lib.t4a_qgrid_int_quantics_to_origcoord(
            self._ptr, quantics_c, len(quantics), out, self._ndims
        )
        check_status(status, "quantics_to_origcoord")
        return [out[i] for i in range(self._ndims)]

    def origcoord_to_grididx(self, coord: Sequence[int]) -> List[int]:
        """Convert original coordinates to grid indices (1-indexed)."""
        lib = get_lib()
        coord_c = ffi.new("int64_t[]", list(coord))
        out = ffi.new("int64_t[]", self._ndims)
        status = lib.t4a_qgrid_int_origcoord_to_grididx(
            self._ptr, coord_c, len(coord), out, self._ndims
        )
        check_status(status, "origcoord_to_grididx")
        return [out[i] for i in range(self._ndims)]

    def grididx_to_origcoord(self, grididx: Sequence[int]) -> List[int]:
        """Convert grid indices (1-indexed) to original coordinates."""
        lib = get_lib()
        grididx_c = ffi.new("int64_t[]", list(grididx))
        out = ffi.new("int64_t[]", self._ndims)
        status = lib.t4a_qgrid_int_grididx_to_origcoord(
            self._ptr, grididx_c, len(grididx), out, self._ndims
        )
        check_status(status, "grididx_to_origcoord")
        return [out[i] for i in range(self._ndims)]

    def grididx_to_quantics(self, grididx: Sequence[int]) -> List[int]:
        """Convert grid indices to quantics indices."""
        lib = get_lib()
        grididx_c = ffi.new("int64_t[]", list(grididx))
        max_sites = sum(self._rs) + self._ndims
        out = ffi.new("int64_t[]", max_sites)
        n_out = ffi.new("size_t*")
        status = lib.t4a_qgrid_int_grididx_to_quantics(
            self._ptr, grididx_c, len(grididx), out, max_sites, n_out
        )
        check_status(status, "grididx_to_quantics")
        return [out[i] for i in range(n_out[0])]

    def quantics_to_grididx(self, quantics: Sequence[int]) -> List[int]:
        """Convert quantics indices to grid indices."""
        lib = get_lib()
        quantics_c = ffi.new("int64_t[]", list(quantics))
        out = ffi.new("int64_t[]", self._ndims)
        n_out = ffi.new("size_t*")
        status = lib.t4a_qgrid_int_quantics_to_grididx(
            self._ptr, quantics_c, len(quantics), out, self._ndims, n_out
        )
        check_status(status, "quantics_to_grididx")
        return [out[i] for i in range(n_out[0])]

    def __repr__(self) -> str:
        return f"InherentDiscreteGrid(ndims={self._ndims}, rs={self._rs})"
